### PR TITLE
fix(gatsby): use faster for loops

### PR DESCRIPTION
**Pull request by KyleAMathews**
Also in source-nodes.ts we looped over all nodes _twice_ so I combined those.

This saves about 1s / 100k nodes.

### CODE REVIEW

File: packages/gatsby-source-drupal/src/gatsby-node.js [hunk 1]
```diff
@@ -191,12 +191,13 @@ exports.sourceNodes = async (
...
-        // Touch nodes so they are not garbage collected by Gatsby.
-        getNodes().forEach(node => {
+        const nodes = getNodes()
+        for (let i = 0; i < nodes.length; i++) {
+          const node = nodes[i]
           if (node.internal.owner === `gatsby-source-drupal`) {
             touchNode(node)
           }
-        })
+        }
```
Comment (wardpeet) [inline packages/gatsby-source-drupal/src/gatsby-node.js:200]: This code path doesn't change anything right? we still loop over all nodes? I rather move it back to `forEach` as it's less code. Or did I miss something?

File: packages/gatsby/src/utils/source-nodes.ts [hunk 2]
```diff
@@ -18,17 +18,9 @@ function discoverPluginsWithoutNodes(
...
-      plugin =>
-        // "Can generate nodes"
-        plugin.nodeAPIs.includes(`sourceNodes`) &&
-        // "Has not generated nodes"
-        !nodeOwnerSet.has(plugin.name)
-    )
-    .map(plugin => plugin.name)
+  for (let i = 0; i < nodes.length; i++) {
+    nodeOwnerSet.add(nodes[i].internal.owner)
+  }
```
Comment (wardpeet) [inline packages/gatsby/src/utils/source-nodes.ts:23]: Same as above right?

File: packages/gatsby/src/utils/source-nodes.ts [hunk 3]
```diff
@@ -49,9 +41,21 @@ function warnForPluginsWithoutNodes(
...
-function getStaleNodes(state: IGatsbyState, nodes: Array<Node>): Array<Node> {
-  return nodes.filter(node => {
+function checkNodes(
+  state: IGatsbyState,
+  nodes: Array<Node>
+): Array<Array<Node> | Array<string>> {
+  const staleNodes: Array<Node> = []
+  const nodeOwnerSet = new Set([`default-site-plugin`])
+  for (let i = 0; i < nodes.length; i++) {
+    const node = nodes[i]
```
Comment (wardpeet) [inline packages/gatsby/src/utils/source-nodes.ts:53]: This can be a `forEach` as well?

File: packages/gatsby/src/utils/source-nodes.ts [hunk 4]
```diff
@@ -70,19 +74,39 @@ function getStaleNodes(state: IGatsbyState, nodes: Array<Node>): Array<Node> {
...
+
+  const noNodesPlugins = state.flattenedPlugins
+    .filter(
+      plugin =>
+        // "Can generate nodes"
+        plugin.nodeAPIs.includes(`sourceNodes`) &&
+        // "Has not generated nodes"
+        !nodeOwnerSet.has(plugin.name)
+    )
+    .map(plugin => plugin.name)
```
Comment (wardpeet) [inline packages/gatsby/src/utils/source-nodes.ts:90]: plugins aren't a big array but while you're optimizing shouldn't this be a forEach loop too so we don't have to loop over it twice?

File: packages/gatsby-source-drupal/src/gatsby-node.js (hunk 1 above)
Comment (KyleAMathews) [reply packages/gatsby-source-drupal/src/gatsby-node.js:200]: These are micro-optimizations. `forEach` is maybe 2x slower than `for` loops — https://leanylabs.com/blog/js-forEach-map-reduce-vs-for-for_of/

Maybe not worth doing 🤷 but node counts can get very large so these microoptimizations start to matter at scale esp for preview/inc builds.

**Comment by KyleAMathews** [comment]
I'm just going to close this. It's not a big improvement and I have other things to do atm.

**Pull request closed without merging**